include/cell.tcc

```diff
@@ -949,3 +949,101 @@ Eigen::VectorXd mpm::Cell<Tdim>::interpolate_nodal_acceleration(
 
   return acceleration;
 }
+
+//! Assign velocity constraint
+//! Constrain directions can take values between 0 and Dim-1
+template <unsigned Tdim>
+bool mpm::Cell<Tdim>::assign_velocity_constraint(unsigned face_id, unsigned dir,
+                                                 double velocity) {
+  bool status = true;
+  try {
+    //! Constrain directions can take values between 0 and Dim * Nphases - 1
+    if (dir >= 0 && dir < Tdim) {
+      if (velocity_constraints_.find(face_id) == velocity_constraints_.end())
+        this->velocity_constraints_.insert(
+            std::make_pair<unsigned, std::vector<std::pair<unsigned, double>>>(
+                static_cast<unsigned>(face_id),
+                std::vector<std::pair<unsigned, double>>()));
+
+      this->velocity_constraints_[face_id].emplace_back(
+          std::make_pair<unsigned, double>(static_cast<unsigned>(dir),
+                                           static_cast<double>(velocity)));
+    } else
+      throw std::runtime_error("Constraint direction is out of bounds");
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+//! Apply velocity constraints
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::apply_velocity_constraints() {
+
+  // TODO
+}
+
+//! Compute all face normal 3d
+template <>
+inline void mpm::Cell<2>::compute_normals() {
+
+  for (unsigned face_id = 0; face_id < 4; ++face_id) {
+    // Get the nodes of the face
+    Eigen::VectorXi indices = element_->face_indices(face_id);
+
+    // Compute the vector to calculate normal (perpendicular)
+    // a = node(0) - node(1)
+    Eigen::Matrix<double, 2, 1> a = (this->nodes_[indices(0)])->coordinates() -
+                                    (this->nodes_[indices(1)])->coordinates();
+
+    // Compute normal and make unit vector
+    // The normal vector n to vector a is defined such that the dot product
+    // between a and n is always 0 In 2D, n(0) = -a(1), n(1) = a(0) Note that
+    // the reverse does not work to produce normal that is positive pointing out
+    // of the element
+    Eigen::Matrix<double, 2, 1> normal_vector;
+    normal_vector(0) = -a(1);
+    normal_vector(1) = a(0);
+    normal_vector = normal_vector.normalized();
+
+    // Store to private variable
+    face_normals_.insert(std::make_pair<unsigned, Eigen::VectorXd>(
+        static_cast<unsigned>(face_id), normal_vector));
+  }
+}
+
+//! Compute all face normal 3d
+template <>
+inline void mpm::Cell<3>::compute_normals() {
+
+  for (unsigned face_id = 0; face_id < 6; ++face_id) {
```

Comment:
Should be generic for any 3D element

Reply:
We will never use triangle right? It's not that generic for 3D and 2D...

Reply:
We may or may not use triangles, but that shouldn't affect how we define this. Get number of faces from the `element` class.